File: Problems/1372. (M) Longest ZigZag Path in Binary Tree/Python3/solution_short.py
class Solution:
  def longestZigZag(self, root: Optional[TreeNode]) -> int:
    # recursive DFS

    def dfs(node, l, r):
      return 0 if node == None else max(
        max(l, r),
          max(dfs(node.left, r+1, 0), dfs(node.right, 0, l+1)))

    return dfs(root, 0, 0)

File: Problems/1372. (M) Longest ZigZag Path in Binary Tree/Java/solution_short.java
class Solution {
  // recursive DFS

  public int longestZigZag(TreeNode root) {
    return dfs(root, 0, 0);
  }

  public int dfs(TreeNode node, int l, int r) {
    return node == null ? 0 : Math.max(
      Math.max(l, r),
      Math.max(dfs(node.left, ++r, 0), dfs(node.right, 0, ++l)));
  }
}
